[AWS] Rework postinstall.sh's backgrounding, add additional error checking, relax pcluster dependencies

lrbison opened the pull request:
Rework nohup, and raise errors as they are encountered

To provide additional confidence that this script has completed
successfully, we `set -euoo posix pipefail` during execution.

Simplify background execution logic to avoid creating temporary
scripts and re-exporting functions.

Improve support for non-pcluster AMIs

Skip install of ACfL compiler if OS is not supported

Commit:
[AWS] Rework nohup, and raise errors as they are encountered

To provide additional confidence that this script has completed
successfully, we `set -euoo posix pipefail` during execution.

Simplify background execution logic to avoid creating temporary
scripts and re-exporting functions.

Signed-off-by: Luke Robison <lrbison@amazon.com>

Commit:
[AWS] Minor logging enhancments

lrbison commented:
@stephenmsachs please take a look 🙏

stephenmsachs commented on AWS/parallelcluster/postinstall.sh:
```diff
@@ -137,6 +137,21 @@ download_yq() {
+    [ -f /etc/os-release ] && . /etc/os-release
+    case "$ID" in
+        amzn|rhel|centos|rocky)
+            yum install -y libssl-dev git curl
```
libssl-dev does not exist for RHEL based (this is probably openssl-devel). Why do we need this on ubuntu?

lrbison replied:
Good catch.  Let me do a more complete test with this change.

I had an issue on ubuntu 22.04 where spack would fail during build on neoverse-v2 platform (no binary cache).

I suspect the spack logic was detecting libssl so using the system package, but then a spack build needed development headers that we didn't have installed, which caused the problem.  I've since lost the log files though.

lrbison replied:
digging this one back up.  I tried to `spack install wrf` and it was unsuccessful:
```
 -   wrf@4.6.1~adios2~chem~netcdf_classic+pnetcdf build_system=generic build_type=dmpar compile_type=em_real nesting=basic patches=68548f6,908c718,f3dd50d arch=linux-ubuntu22.04-neoverse_v1 %gcc@12.4.0 
 -       ^libtirpc@1.3.3 build_system=autotools arch=linux-ubuntu22.04-neoverse_v1 %gcc@12.4.0
 -           ^krb5@1.21.3+shared build_system=autotools arch=linux-ubuntu22.04-neoverse_v1 %gcc@12.4.0                   
[e]              ^openssl@3.0.2~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu22.04-neoverse_v1 %gcc@12.4.0
```

build of krb5 fails with:
```
In file included from pkinit_crypto_openssl.h:36,
                 from pkinit_crypto_openssl.c:33:
/home/ubuntu/spack/opt/spack/linux-amzn2-aarch64/gcc-7.3.1/gcc-12.4.0-v6wxye6ijzrxnzxftcwnpu3psohsjl2b/bin/../lib/gcc/aarch64-unknown-linux-gnu/12.4.0/include-fixed/openssl/bn.h:831:6: error: conflicting types for 'ERR_load_BN_strings'; have 'void(void)'
  831 | void ERR_load_BN_strings(void);
      |      ^~~~~~~~~~~~~~~~~~~
```

krb5 needs the headers.

stephenmsachs replied:
We can add them here to cover this case, but I am afraid this is no the long term solution. For once the spack installation is available on more than one node and this seems to be a mismatch between gcc and libssl.

lrbison replied:
Are you are suggesting we prevent openssl from being loaded externally so that spack can properly install a version with headers?  Right now spack won't attempt to install openssl because it finds it as an external

lrbison replied:
discussed this offline.  It seems the issue is likely the fixedheaders described by GCC here: https://gcc.gnu.org/onlinedocs/gcc-7.5.0/gcc/Fixed-Headers.html.  The system-header for whatever openssl AL2 had was packaged into the binary cache and distributed into my ubuntu system.  Problems follow from there.  I will file a bug separately.

stephenmsachs replied:
This can be resolved by https://github.com/spack/spack/pull/49641

stephenmsachs commented on AWS/parallelcluster/postinstall.sh:
```diff
@@ -523,6 +552,7 @@ echo "$(declare -pf)
 if ${install_in_foreground}; then
     bash ${tmpfile}
 else
+    echo "Continuing in background.  Watch /var/log/spack-postinstall.log"
```
Great catch. Should have added this a long time ago.

stephenmsachs commented on AWS/parallelcluster/postinstall.sh:
```diff
@@ -352,6 +368,17 @@ setup_pcluster_buildcache_stack() {
     [ -n "${LIBFABRIC_VERSION}" ] && yq -i ".packages.libfabric.externals[0].spec = \"libfabric@${LIBFABRIC_VERSION}\"" "${SPACK_ROOT}"/etc/spack/packages.yaml
 }
 
+remove_missing_packages() {
```
I think it would be easier to read if we just inline this [here](https://github.com/spack/spack-configs/blob/012c81bb1c1b29388f1d2ee33c47aa905647bd82/AWS/parallelcluster/postinstall.sh#L349)

lrbison replied:
Leaving it as I had it here, because I want to remove them in both the --buildcache and --no-buildcache paths.

stephenmsachs commented on AWS/parallelcluster/postinstall.sh:
```diff
@@ -137,6 +137,21 @@ download_yq() {
+            ;;
+        ubuntu)
+            apt-get update
+            apt-get install -y libssl-dev git curl
```
Matt found that ubuntu apparently [sometimes ships without unzip](https://github.com/aws-samples/aws-hpc-recipes/blob/b2fb6736e74f26b9174a1febfeb955793e5c2a99/recipes/pcs/hpc_ready_ami/assets/scripts/install-spack.sh#L84). We should add this here as well (ignore the python3-pip, that's an older version of this file).

lrbison replied:
I still need to address this.

lrbison replied:
addressed.

stephenmsachs reviewed (changes requested):
Looks good to me except for the libssl-dev stuff.

Commit:
[AWS] Skip install of ACfL compiler if OS is not supported

Commit:
[AWS] Improve support for non-pcluster AMIs

lrbison commented:
This has turned into a larger change.  My motivations are two-fold:

  1. I want to be able to use this script even on a complete single node without pcluster or PCS to install spack with binary cache support on AWS platforms.  To this end we need to remove some dependency on slurm and libfabric which we previously assumed.
  2. I want to be able to confidently test the script has completed successfully.  This helps me test on many OSs using scripts.

I've tested this version on the following non-pcluster AMIs:
 - Ubuntu 20.04, 22.04, and 24.04
 - Rocky 9.5
 - Amazon Linux 2 and 2023.
 
However I've recently realized my testing scripts pre-installed more packages than I've realized, so I'll need to address that, and I still need to test on pcluster amis to double-check I haven't broken anything.

stephenmsachs commented on AWS/parallelcluster/postinstall.sh:
```diff
@@ -465,18 +569,18 @@ setup_mirrors() {
     fi
     # Add older specific target mirrors if they exist
-    mirror_url="https://binaries.spack.io/develop/aws-pcluster-$(target | sed -e 's?_avx512??1')"
+    mirror_url="https://binaries.spack.io/${SPACK_BRANCH:-develop}/aws-pcluster-$(target | sed -e 's?_avx512??1')"
```
Same as above applies here. Furthermore, `aws-pcluster-$(target | sed -e 's?_avx512??1')` stacks are no longer being built so this only exists in `develop` (which IIRC holds all the old builds).

lrbison replied:
changed this back to just `develop`

stephenmsachs commented on AWS/parallelcluster/postinstall.sh:
```diff
@@ -465,18 +569,18 @@ setup_mirrors() {
+    mirror_url="https://binaries.spack.io/${SPACK_BRANCH:-develop}/aws-pcluster-$(target | sed -e 's?_avx512??1')"
     if curl -fIsLo /dev/null "${mirror_url}/build_cache/index.json"; then
-        spack mirror add --scope site "aws-pcluster-legacy" "${mirror_url}"
+        spack mirror list | grep -q $mirror_url || \
```
This is good. We should add this everywhere since `set -e` is set. However, this does not work with `set -o pipefail`.

stephenmsachs commented on AWS/parallelcluster/postinstall.sh:
```diff
@@ -465,18 +569,18 @@ setup_mirrors() {
+            spack mirror add --scope site "aws-pcluster-legacy" "${mirror_url}"
     fi
     spack mirror list | grep -qE '\[.*b.*\]' && spack buildcache keys -it
+    return 0
```
Again, I don't understand why.

lrbison replied:
Although I didn't observe this failure path, I added it defensively in case the `grep` failed to match thus returning non-zero and passing it up as the function's return value.

stephenmsachs reviewed (changes requested):
Thanks for your help. I like the changes you did, especially the new `-fg` option. I am not sure if all those bash options are necessary and help reproducibility. And we will have to clean up in which cases the `export` statements are still necessary, but that's not essential.

lrbison commented on AWS/parallelcluster/postinstall.sh:
```diff
@@ -350,6 +421,18 @@ setup_pcluster_buildcache_stack() {
     [ -n "${SLURM_VERSION}" ] && yq -i ".packages.slurm.externals[0].spec = \"slurm@${SLURM_VERSION} +pmix\"" "${SPACK_ROOT}"/etc/spack/packages.yaml
     [ -d "${SLURM_ROOT}" ] && yq -i ".packages.slurm.externals[0].prefix = \"${SLURM_ROOT}\"" "${SPACK_ROOT}"/etc/spack/packages.yaml
     [ -n "${LIBFABRIC_VERSION}" ] && yq -i ".packages.libfabric.externals[0].spec = \"libfabric@${LIBFABRIC_VERSION}\"" "${SPACK_ROOT}"/etc/spack/packages.yaml
+    return 0
```
This was a fun one.  The last statement was a test condition that can (and did) set $? to non-zero.  This is allowed in `set -e` mode as long as the statement was used in a test condition, so it didn't immediately return an error.  However once the function exits it inherits $? as the last statement executed and returns that, which IS an error to `set -e` mode because the function itself is not used in a conditional.

lrbison commented on AWS/parallelcluster/postinstall.sh:
```diff
@@ -362,7 +445,11 @@ setup_spack() {
     for compiler in $(spack compiler list | grep -v '-' |grep -v '=>' | xargs); do
-        spack compiler info --scope=site ${compiler} 2>/dev/null | grep -E 'cc =|cxx =' | grep -q " None" && spack compiler rm --scope=site ${compiler}
+        compiler_ok=no
+        spack compiler info --scope=site ${compiler} 2>/dev/null | grep -E 'cc =|cxx =' | grep -q " None" || compiler_ok=yes
```
the intent is to find compilers that have either `cc = None` or `cxx = None` and remove them.  However for good compilers neither of those patterns will match, and grep will return 1.  This causes the ` && spack compiler rm ...` portion not to run (as desired), but now with `-e` the whole statement has a return value of 1, and this causes the postinstall.sh script to exit with error condition.

The new version guards the return value with the assignment, and tests the assignment in a separate conditional.

lrbison commented on AWS/parallelcluster/postinstall.sh:
```diff
@@ -465,18 +569,18 @@ setup_mirrors() {
-        else
-            spack mirror add --scope site "aws-pcluster-$(stack_arch)" "https://binaries.spack.io/develop/aws-pcluster-$(stack_arch)"
-        fi
+        mirror_url="https://binaries.spack.io/${SPACK_BRANCH:-develop}/aws-pcluster-$(stack_arch)"
```
I see.  So I tested it and it seems that I can add a mirror with a bad URL and spack does not complain.  So to be specific it may fail doing something at a later time (I couldn't make my example fail) or it might just silently ignore it.

But your point is valid: if I close a different version of spack, perhaps I should still try to use the develop mirror?

How about we add both the develop and the branch mirror.  If the user chooses to install with some feature branch, then we'll get develop binaries, and if they install with some old branch, we'll get the old binaries (and none of the develop ones unless they have a matching hash).

What do you think?

lrbison commented on AWS/parallelcluster/postinstall.sh:
```diff
@@ -465,18 +569,18 @@ setup_mirrors() {
-            spack mirror add --scope site "aws-pcluster-$(stack_arch)" "https://binaries.spack.io/develop/aws-pcluster-$(stack_arch)"
-        fi
+        mirror_url="https://binaries.spack.io/${SPACK_BRANCH:-develop}/aws-pcluster-$(stack_arch)"
+        spack mirror list | grep -q $mirror_url || \
```
no, the || prevents the failure.  Anyways the pipefail only effects if `spack mirror list` is failing, the failure we need to guard against with || is from the grep.

lrbison commented on AWS/parallelcluster/postinstall.sh:
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-set -e
+set -euoo posix pipefail
```
responses in order:
 - `if [ -z "${FOO}" ]` does not work, but `if [ -z "${FOO:-}" ]` does.  However I think I caught them all.  Do you see examples where I missed some?
 - https://www.gnu.org/software/bash/manual/html_node/Bash-POSIX-Mode.html I inherited this habit from my team.  I don't have strong opinions.
 - re pipefail: I think we have a different understanding here, perhaps we can talk offline?

lrbison commented on AWS/parallelcluster/postinstall.sh:
```diff
@@ -137,6 +178,22 @@ download_yq() {
+    [ -f /etc/os-release ] && . /etc/os-release
+    case "$ID" in
+        amzn|rhel|centos|rocky)
+            yum install -y git gcc gcc-c++ gcc-gfortran make cmake glibc-devel
```
On the fortran point: I recall some problems where spack would find a c/c++ compiler and assume fortran was installed as well.  Perhaps these have been fixed.  I assumed it was rather painless to install all 3 from OS.

It seems you are concerned we might end up with different versions of the compilers.  I am a little wary of depending on a package that we didn't mark for install... although I suppose it's safe since we know c++ depends on gcc.

Something did depend on them... I forgot what.  On the debian side build-essential brings them in.

lrbison commented on AWS/parallelcluster/postinstall.sh:
```diff
@@ -211,6 +269,7 @@ cleanup() {
 download_spack() {
     if [ -z "${SPACK_ROOT}" ]
     then
+        echo "CLONING: $SPACK_ROOT and $install_path"
```
debug print slipped in, but I do think it could be helpful.  Reworking it.

stephenmsachs reviewed (commented):
I would still remove the pipefail since it adds more than it helps, e.g. the pipe in line 345 is allowed to fail safely.

lrbison commented:
@stephenmsachs I believe I have addressed your recommended changes:
 - removed pipefail
 -  No libssl stuff (tracked in https://github.com/spack/spack/issues/49640)

Are we ready for merge?
